homeassistant/components/firmata/__init__.py

```diff
@@ -0,0 +1,142 @@
+from copy import deepcopy
+import logging
+
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_NAME, EVENT_HOMEASSISTANT_STOP
+from homeassistant.helpers import config_validation as cv
+
+from . import config_flow  # noqa: F401
+from .board import FirmataBoard
+from .const import (
+    CONF_ARDUINO_INSTANCE_ID,
+    CONF_ARDUINO_WAIT,
+    CONF_BINARY_SENSORS,
+    CONF_INITIAL_STATE,
+    CONF_NEGATE_STATE,
+    CONF_PIN,
+    CONF_PIN_MODE,
+    CONF_SAMPLING_INTERVAL,
+    CONF_SERIAL_BAUD_RATE,
+    CONF_SERIAL_PORT,
+    CONF_SLEEP_TUNE,
+    CONF_SWITCHES,
+    DOMAIN,
+    PIN_MODE_INPUT,
+    PIN_MODE_OUTPUT,
+    PIN_MODE_PULLUP,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_CONFIGS = "board_configs"
+
+ANALOG_PIN_SCHEMA = vol.All(cv.string, vol.Match(r"^A[0-9]+$"))
+
+SWITCH_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_NAME): cv.string,
+        vol.Required(CONF_PIN): vol.Any(cv.positive_int, ANALOG_PIN_SCHEMA),
+        # will be analog mode in future too
+        vol.Required(CONF_PIN_MODE): PIN_MODE_OUTPUT,
+        vol.Optional(CONF_INITIAL_STATE, default=False): cv.boolean,
+        vol.Optional(CONF_NEGATE_STATE, default=False): cv.boolean,
+    },
+    required=True,
+    extra=vol.ALLOW_EXTRA,
+)
+
+BINARY_SENSOR_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_NAME): cv.string,
+        vol.Required(CONF_PIN): vol.Any(cv.positive_int, ANALOG_PIN_SCHEMA),
+        # will be analog mode in future too
+        vol.Required(CONF_PIN_MODE): vol.Any(PIN_MODE_INPUT, PIN_MODE_PULLUP),
+        vol.Optional(CONF_NEGATE_STATE, default=False): cv.boolean,
+    },
+    required=True,
+    extra=vol.ALLOW_EXTRA,
+)
+
+BOARD_CONFIG_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_SERIAL_PORT): cv.string,
+        vol.Optional(CONF_SERIAL_BAUD_RATE): cv.positive_int,
+        vol.Optional(CONF_ARDUINO_INSTANCE_ID): cv.positive_int,
+        vol.Optional(CONF_ARDUINO_WAIT): cv.positive_int,
+        vol.Optional(CONF_SLEEP_TUNE): vol.All(
+            vol.Coerce(float), vol.Range(min=0.0001)
+        ),
+        vol.Optional(CONF_SAMPLING_INTERVAL): cv.positive_int,
+        vol.Optional(CONF_SWITCHES): [SWITCH_SCHEMA],
+        vol.Optional(CONF_BINARY_SENSORS): [BINARY_SENSOR_SCHEMA],
+    },
+    required=True,
+    extra=vol.ALLOW_EXTRA,
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.All(cv.ensure_list, [BOARD_CONFIG_SCHEMA])}, extra=vol.ALLOW_EXTRA
+)
+
+
+async def async_setup(hass, config):
+    """Set up the Firmata domain."""
+    # Delete all current entries and re-add them to update them
+    if hass.config_entries.async_entries(DOMAIN) and DOMAIN in config:
+        for entry in hass.config_entries.async_entries(DOMAIN):
+            await hass.config_entries.async_remove(entry.entry_id)
+    if DOMAIN in config:
+        for board in config[DOMAIN]:
+            firmata_config = deepcopy(board)
```

Comment:
Why deepcopy it ?

Reply:
I wrote this a while ago now, so it is hard to remember, but it was intentional. I believe it was to break the references. I will look into it again.

Reply:
deepcopy should only be done if you're planning on mutating the dictionary inside the config flow.